core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala

```diff
@@ -118,6 +118,7 @@ private[spark] class PipedRDD[T: ClassTag](
 
     val proc = pb.start()
     val env = SparkEnv.get
+    @volatile var exception: Option[Throwable] = None
```

Comment:
Let's use an AtomicReference here, and we can also get rid of the Option (just use null since it is localized scope)